ntoskrnl/mm/section.c

```diff
@@ -4876,7 +4876,8 @@ MmPurgeSegment(
     }
 
     /* Find byte offset of the page to start */
-    PurgeStart.QuadPart = PAGE_ROUND_DOWN(PurgeStart.QuadPart);
+    PurgeStart.QuadPart >>= PAGE_SHIFT;
+    PurgeStart.QuadPart <<= PAGE_SHIFT;
```

Comment:
```suggestion
    PurgeStart.LowPart = PAGE_ROUND_DOWN(PurgeStart.LowPart);
// or
    PurgeStart.LowPart &= ~(PAGE_SIZE - 1);
```

Reply:
Offset can be 64-bit so I think using only `LowPart` is not going to work.

Reply:
Why not? `PAGE_SIZE` is 32-bit (even 16-bit).

Reply:
I mean `PurgeStart` itself which can be a 64-bit value.

Reply:
Yeah, but rounding down does not affect `PurgeStart.HighPart`, does it?

Reply:
I'm not sure if that is true for all values or not, but your suggestion looks hacky to me.

Reply:
> I'm not sure if that is true for all values or not

Well, it should...

> your suggestion looks hacky to me.

A bit, maybe.
Yet, this is what some existing codes use, especially:
https://git.reactos.org/?p=reactos.git&a=search&h=HEAD&st=grep&s=%5C.LowPart+%5C%26%5C%3D+%7E%5C%28PAGE_SIZE+*-+*1%5C%29%3B&sr=1

Reply:
OK. It should be fine then.
Since the fastfat driver is from MS, it's funny that they use a somewhat hacky code instead of adding proper macros for it.